### PR TITLE
Diverging tyvars

### PR DESCRIPTION
**Pull request by lovishpuri**
Rebased with master and resolved conflicts.

-- Copied from #85021 --
Previously, the never type will be replaced by a diverging type variable (generally) only if some coercion occurs. It can cause inconsistent behaviors like
```rust
return.foo();      //~ ERROR no method named `foo` found for type `!`
{ return }.foo();  //~ ERROR type annotations needed
```
```rust
let a = (return, );          // The type is `(!, )`.
let a = ({ return }, );      // The type is `(_, )`.
let a: (_, ) = (return, );   // The type is `(_, )`.
```

With the first commit, the never type will be replaced by a diverging type variable just at the end of the type check for every expression, even if no coercion occurs. Thus the problems above get solved and the consistency should be improved.

Then, another problem is we'll issue too many "type annotations needed". They are issued when types must be known at some point but the resolution failed, even if the type variables are just some diverging ones. A typical example is `{ return }.foo()`.

With the second commit, the information about diverging is recorded in the unification table, so that we can check whether performing the fallback affects other non-diverging type variables. If it doesn't, we will safely perform the fallback and we won't issue "type annotations needed" anymore.

As a result, "type annotations needed" will be issued for
```rust
let a = return;
{ if true { a } else { return } }.foo();
```
but not for
```rust
let a: ! = return;
{ if true { a } else { return } }.foo();
```

Discussed at https://rust-lang.zulipchat.com/#narrow/stream/259160-t-lang.2Fproject-never-type/topic/Never.20type.20in.20tuple .

### CODE REVIEW

File: src/test/ui/issues/issue-13847.rs [hunk 1]
```diff
@@ -1,3 +1,3 @@
 fn main() {
-    return.is_failure //~ ERROR no field `is_failure` on type `!`
+    return.is_failure //~ ERROR no field `is_failure` on type `()`
```
Comment (jackh726) [inline src/test/ui/issues/issue-13847.rs:2]: Haven't looked much at the implementation here, but this change alone signals something is probably not correct with this PR; the type of `return` *should* be `!`, not `()`

**Review by jackh726 — COMMENTED** [review]
Can you give some background here? What is this PR trying to solve?

**Comment by jackh726** [comment]
Is this just a rebase of #85021?

**Comment by jackh726** [comment]
Going to reassign to Niko, since he is assigned to the other PR r? @nikomatsakis

**Comment by nikomatsakis** [comment]
cc @Mark-Simulacrum we should discuss this pr =) I can't quite remember what it does

**Comment by Mark-Simulacrum** [comment]
Apologies for the long wait -- taking a look at this PR, I think in broad strokes the goal is not yet clear to me. While there is some "inconsistency" in the handling of `return.foo()` vs. `{ return }.foo()`, I'm not sure that we should actually try to fix it.

It is true that we try to infer the type of the block (vs. a literal return expression), which causes an inference failure rather than the method resolution failure, but I'm not sure that this is actually a problem in practice? Such code is pretty rare, I'd expect, and introducing more code that "cares" about never type in particular seems unfortunate.

The second change which affects whether we indicate inference failed (i.e., type annotations needed) also seems somewhat similar in practice -- do we have real-world(ish) code examples of that change improving things?

It might also be a rebase error -- not clear if the diff github showing is accurate -- but the changes to the MIR-opt tests do concern me a little. Ideally I think we'd not see changes to compiling code based on changes to inference, since that means we're at least potentially altering the runtime behavior.

Can you say a little more about what motivated this patch? Is it just to try to close out the particular issue around whether never-typed expressions sometimes get inference variables and sometimes don't?

**Comment by JohnCSimon** [comment]
Ping from triage: 
@lovishpuri Can you please address the merge conflicts and comments from the reviewer? Thanks.

**Comment by JohnCSimon** [comment]
Ping from triage: 
@lovishpuri 
I'm closing this due to inactivity, Please reopen when you are ready to continue with this. Thank you.

@rustbot label: +S-inactive

**Pull request closed without merging**